Update Substrate PalletXCM and Send XC-20s Overview (#1087)
moonbeam-foundation/moonbeam-docs: builders/interoperability/xcm/xc20/send-xc20s/overview.md

Moonbeam currently uses remote transfers for XC-20 transfers.

This page covers the fundamentals of XCM-based remote transfers. To learn how to perform XC-20 transfers, refer to the the [XC-20 transfers via the Substrate API](/builders/interoperability/xcm/xc20/send-xc20s/xcm-pallet/){target=\_blank} guide.

## XCM Instructions for Asset Transfers {: #xcm-instructions-for-asset-transfers }

The XCM Pallet and Precompile abstract much of the complexity involved in cross-chain asset transfers, automatically constructing the necessary XCM messages. Nevertheless, having a basic understanding of the underlying instructions can be useful. 

For reference, you can find the Polkadot XCM Pallet extrinsics for sending XC-20s in the [Using the Polkadot XCM Pallet To Send XC-20s guide](/builders/interoperability/xcm/xc20/send-xc20s/xcm-pallet/){target=_blank}.

The instructions in each XCM transfer vary depending on the asset and the transfer route. For example, returning a native token like xcDOT to its reserve chain (from Moonbeam to Polkadot) differs from sending DOT from Polkadot to Moonbeam. Below are examples of the instructions commonly involved in these token transfers.

> ### Instructions to Transfer a Reserve Asset from the Reserve Chain {: #instructions-to-transfer-a-reserve-asset-from-the-reserve-chain }

Reviewer: Careful with changing the header tag

--8<-- 'text/builders/interoperability/xcm/xc20/send-xc20s/overview/DOT-to-xcDOT-instructions.md'

> This process invokes TransferReserveAsset with `assets`, `dest`, and `xcm`parameters. Within the `xcm` parameter, you typically specify the `BuyExecution` and `DepositAsset` instructions. As shown in the [`TransferReserveAsset` instruction](https://github.com/paritytech/polkadot-sdk/blob/{{ polkadot_sdk }}/polkadot/xcm/xcm-executor/src/lib.rs#L630){target=\_blank}, the flow also includes `ReserveAssetDeposited` and `ClearOrigin` to complete the transfer.

Reviewer 2: ```suggestion
This process invokes `TransferReserveAsset` with `assets`, `dest`, and `xcm`parameters. Within the `xcm` parameter, you typically specify the `BuyExecution` and `DepositAsset` instructions. As shown in the [`TransferReserveAsset` instruction](https://github.com/paritytech/polkadot-sdk/blob/{{ polkadot_sdk }}/polkadot/xcm/xcm-executor/src/lib.rs#L630){target=\_blank}, the flow also includes `ReserveAssetDeposited` and `ClearOrigin` to complete the transfer.
```

For more information on constructing an XCM message for asset transfers, such as DOT to Moonbeam, refer to the [Polkadot XCM Pallet guide](/builders/interoperability/xcm/xc20/send-xc20s/xcm-pallet/){target=\_blank}.

### Instructions to Transfer a Reserve Asset back to the Reserve Chain {: #transfer-native-to-origin }

--8<-- 'text/builders/interoperability/xcm/xc20/send-xc20s/overview/xcDOT-to-DOT-instructions.md'

> Steps 3 through 6 are automatically triggered by the `InitiateReserveWithdraw` instruction (Step 2) and execute on Polkadot. Once `InitiateReserveWithdraw` is invoked on Moonbeam, the assembled XCM message instructs Polkadot to run those final instructions, completing the cross-chain transfer. In other words, while Moonbeam constructs the XCM instructions behind the scenes, they ultimately execute on Polkadot to complete the asset’s return to its reserve chain.

Reviewer 2: ```suggestion
Steps 3 through 6 are automatically triggered by the `InitiateReserveWithdraw` instruction (step 2) and execute on Polkadot. Once `InitiateReserveWithdraw` is invoked on Moonbeam, the assembled XCM message instructs Polkadot to run those final instructions, completing the cross-chain transfer. In other words, while Moonbeam constructs the XCM instructions behind the scenes, they ultimately execute on Polkadot to complete the asset’s return to its reserve chain.
```

## XCM Instructions for Asset Transfers {: #xcm-instructions-for-asset-transfers }

The `X-Tokens` Pallet and `XCM` Precompile simplify cross-chain asset transfers by abstracting away the process of building the underlying XCM message. This guide provides insight into the underlying instructions used by XCM to transfer assets cross-chain.

For reference, the XCM extrinsics used by the Polkadot XCM Pallet are defined at the top of the [Using the Polkadot XCM Pallet To Send XC-20s guide](/builders/interoperability/xcm/xc20/send-xc20s/xcm-pallet/){target=_blank}.

The specific instructions in each XCM depend on the asset being transferred and the route taken. For example, sending a native asset back to its reserve chain (for instance, xcDOT from Moonbeam back to Polkadot) differs from sending it from its origin to a new chain (DOT from Polkadot to Moonbeam).

Examples of the XCM instructions typically involved in token transfers are included below.

### Instructions to Transfer a Reserve Asset from the Reserve Chain {: #transfer-native-from-origin }

--8<-- 'text/builders/interoperability/xcm/xc20/send-xc20s/overview/DOT-to-xcDOT-instructions.md'

> For more information on constructing an XCM message for self-reserve (native) assets—such as DOT to Moonbeam—refer to the [Polkadot XCM Pallet guide](https://docs.moonbeam.network/builders/interoperability/xcm/xc20/send-xc20s/xcm-pallet/){target=_blank}.

Reviewer 2: ```suggestion
For more information on constructing an XCM message for self-reserve (native) asset transfers—such as DOT to Moonbeam—refer to the [Polkadot XCM Pallet guide](/builders/interoperability/xcm/xc20/send-xc20s/xcm-pallet/){target=\_blank}.
```

Can you please double-check the links and add back the backslashes?

> This process calls `TransferReserveAsset` with `assets`, `dest`, and `xcm` parameters. Within `xcm`, the typical instructions are `BuyExecution` and `DepositAsset`. The [`TransferReserveAsset` instruction](https://github.com/paritytech/polkadot-sdk/blob/{{ polkadot_sdk }}/polkadot/xcm/xcm-executor/src/lib.rs#L671){target=_blank} on GitHub includes `ReserveAssetDeposited` and `ClearOrigin` alongside `BuyExecution` and `DepositAsset` to finalize the transfer.

Reviewer 2: How can they track back where the `xcm` parameter is assembled? The whole point of having these links here are so that devs can make sense of it and see for themselves how the messages are built.

The link, which points to line 671, is referencing a line in the `ReceiveTeleportedAsset` function. So that needs to be updated. It should be 646, I believe. But that alone is confusing. There used to be a paragraph above this one that referenced calling the `TransferReserveAsset` function, including the assembly of the `xcm` instructions, but it's no longer here because it referenced X-Tokens, but there should be an equivalent elsewhere now.

I feel like as it is now, yes, it does make sense, but we're not connecting all of the dots.
